### PR TITLE
Fixes #25395: Fusion inventory fails to report software with exotic names -  some entries randomly disappear

### PR DESCRIPTION
**Pull request by peckpeck**
https://issues.rudder.io/issues/25395

### CODE REVIEW

File: rudder-agent/SOURCES/patches/fusioninventory/25395-windows-utf8.patch [hunk 1]
```diff
@@ -0,0 +1,813 @@
...
+         () : getCpusFromDmidecode();
+ 
+-    # the CPU description in WMI is false, we use the registry instead
+-    my $registryInfos = getRegistryKey(
+-        path   => "HKEY_LOCAL_MACHINE/Hardware/Description/System/CentralProcessor",
+-        wmiopts => { # Only used for remote WMI optimization
+-            values  => [ qw/Identifier ProcessorNameString VendorIdentifier/ ]
+-        }
++    my $registryInfos = getNewRegistryAll(
++        logger => $params{logger},
```
Comment (Fdall) [inline rudder-agent/SOURCES/patches/fusioninventory/25395-windows-utf8.patch:101]: Why is the logger only passed here ? What does it log ?
If the logging is automatic, I think that we should pass it whenever we use the getNewRegistryAll function in our patch

File: rudder-agent/SOURCES/patches/fusioninventory/25395-windows-utf8.patch [hunk 2]
```diff
@@ -0,0 +1,813 @@
...
+ 
+     foreach my $versionKey (keys %{$key}) {
+-        my $registrationKey = $key->{$versionKey}->{'Registration/'};
++        my $registrationKey = $key->{$versionKey}->{Registration};
+         next unless $registrationKey;
+ 
+         foreach my $uuidKey (keys %{$registrationKey}) {
+diff --color -ruN fusion-source.orig/lib/FusionInventory/Agent/Task/Inventory/Win32/License.pm.orig fusion-source/lib/FusionInventory/Agent/Task/Inventory/Win32/License.pm.orig
+--- fusion-source.orig/lib/FusionInventory/Agent/Task/Inventory/Win32/License.pm.orig	1970-01-01 01:00:00.000000000 +0100
++++ fusion-source/lib/FusionInventory/Agent/Task/Inventory/Win32/License.pm.orig	2020-08-04 09:36:42.000000000 +0200
```
Comment (ncharles) [inline rudder-agent/SOURCES/patches/fusioninventory/25395-windows-utf8.patch:178]: why is there a .pm.orig file ?

File: rudder-agent/SOURCES/patches/fusioninventory/25395-windows-utf8.patch [hunk 3]
```diff
@@ -0,0 +1,813 @@
...
++
++    for (my $i=0;$i<10;$i++) {
++		if (RegOpenKeyExW($root, encode("UTF-16LE", $path), 0, KEY_ENUMERATE_SUB_KEYS, $swKey)) {
++			last;
++		}
++		# sometimes windows fails to find it own root registry, who knows why ...
++		if (regLastError() != 2) {
++			last;
++		}
++		sleep(0.1);
```
Comment (ncharles) [inline rudder-agent/SOURCES/patches/fusioninventory/25395-windows-utf8.patch:702]: maybe you could increment a bit the sleep ? like to 0.2 ?

File: rudder-agent/SOURCES/patches/fusioninventory/25395-windows-utf8.patch [hunk 4]
```diff
@@ -0,0 +1,813 @@
...
++    my $keyLastWrite;
++
++    my $path = $params{path};
++    $path =~ s/\//\\\\/g;
++
++    my $root = _getNewRegistryRoot(
++        root => $params{root}
++    );
++
++    for (my $i=0;$i<10;$i++) {
```
Comment (ncharles) [inline rudder-agent/SOURCES/patches/fusioninventory/25395-windows-utf8.patch:694]: you should probably increase this one to 20 ?

**Comment by peckpeck** [comment]
PR updated with a new commit

**Comment by peckpeck** [comment]
ping  @Fdall @ncharles

**Review by ncharles — APPROVED** [review]

**Pull request merged**